src/core_plugins/kibana/public/dashboard/directives/dashboard_panel.js

```diff
@@ -69,52 +52,124 @@ uiModules
       /**
        * @type {boolean}
        */
-      isExpanded: '='
+      isExpanded: '=',
+      /**
+       * Returns a click handler for a visualization.
+       * @type {function}
+       */
+      getVisClickHandler: '&',
+      /**
+       * Returns a brush event handler for a visualization.
+       * @type {function}
+       */
+      getVisBrushHandler: '&',
+      /**
+       * Call when changes should be propagated to the url and thus saved in state.
+       * @type {function}
+       */
+      saveState: '&'
     },
     link: function ($scope, element) {
       if (!$scope.panel.id || !$scope.panel.type) return;
 
-      loadPanel($scope.panel, $scope)
-      .then(function (panelConfig) {
-        // These could be done in loadPanel, putting them here to make them more explicit
-        $scope.savedObj = panelConfig.savedObj;
-        $scope.editUrl = panelConfig.editUrl;
+      /**
+       * Retrieves the saved object represented by the panel and returns it, along with the appropriate
+       * edit Url.
+       * @returns {Promise.<{savedObj: SavedObject, editUrl: String}>}
+       */
+      function loadSavedObject() {
+        /**
+         * The appropriate loader for the panel type.
+         * @type {SavedObjectLoader}
+         */
+        let loader;
+        /**
+         * A custom function that will run after loading the saved object, specific to the type.
+         * @type {function}
+         */
+        let postLoadInit;
+
+        if ($scope.panel.type === 'search') {
```

Comment:
The problem I have with this explicit conditional is that it does not scale well to new panel types:
* Previously, when adding a new panel type, a new loader would just have to be created and added to the declarative `panelTypes` mapping in `loadPanelProvider`. The panel code itself would not have to be changed.
* This conditional though would grow with each new panel type and probably become a lot harder to read.